src/CMakeLists.txt

```diff
@@ -14,21 +14,23 @@ if (NOT BUILD_SHARED_LIBS)
 endif ()
 
 # Set the (shared) library version
-set(Halide_VERSION_OVERRIDE "${Halide_VERSION}"
+set(Halide_VERSION_OVERRIDE "<default>"
     CACHE STRING "VERSION to set for custom Halide packaging")
 mark_as_advanced(Halide_VERSION_OVERRIDE)
 
-if (Halide_VERSION_OVERRIDE)
```

Comment:
Can't you just do `if (DEFINED Halide_VERSION_OVERRIDE)`?

Reply:
No, because, as a cache variable, it is always defined.

I guess it might suffice to initialize it to the default and use it unconditionally?

Reply:
Ah, wait this patch is still wrong because I need a way to include no information at all.

Reply:
I think this is better now, @LebedevRI

Reply:
I think this is less messy, but there's an issue.
I currently do pass `-DHalide_VERSION_OVERRIDE=""`,
in order to end up with `libhalide<major>.so.<soversion>` + `libhalide<major>.so` (symlink to former)...
I would really suggest `if(DEFINED <???>)`, but in the end it doesn't matter.

Reply:
If you set `VERSION` to the empty string and set `SOVERSION` to something meaningful, you get:

```
libHalide..dylib  # real lib
libHalide.19.dylib -> libHalide.dylib
libHalide.dylib -> libHalide.19.dylib
```

That first entry, with the truly empty version component, is insane. If you leave `VERSION` _undefined_, however, you get what you described:

```
libHalide.19.dylib  # real lib
libHalide.dylib -> libHalide.19.dylib
```

For pip packaging I need to _just_ keep `libHalide.dylib` with no symlinks. This is because pip wheels are ZIP formatted so the symlinks triple the size of the archive. I achieve this by leaving both `VERSION` and `SOVERSION` undefined, which is what I have in this PR set the meaning of `-DHalide_[SO]VERSION_OVERRIDE=""` to be. Actually setting them to the empty string never seems necessary.

Reply:
As i've said, for all i care, this patch is unnessesary, i will still have to patch stuff on the packaging side.

> If you set `VERSION` to the _empty string_ and set `SOVERSION` to something meaningful, you get:
> 
> ```
> libHalide..dylib  # real lib
> libHalide.19.dylib -> libHalide.dylib
> libHalide.dylib -> libHalide.19.dylib
> ```
> 
> That first entry, with the truly empty version component, is insane.

I agree it is, but that is not what i get here. That insanity might be osx-specific.

> If you leave `VERSION` _undefined_, however, you get what you described:

No, i don't. I need major version to be a suffix of the name, without `.`-separator.

Reply:
> No, i don't. I need major version to be a suffix of the name, without `.`-separator.

What you're describing here would be something like `CMAKE_RELEASE_POSTFIX`, no?

Reply:
With this patch, you can set:

```
-DCMAKE_RELEASE_POSTFIX=19 -DHalide_VERSION_OVERRIDE="" -DHalide_SOVERSION_OVERRIDE=0
```

I get `libHalide19.0.so` and a symlink to it called `libHalide19.so`.

Reply:
Maybe.

Let me put it another way: why must this cmakefile refuse to take `""` as valid version/soversion override?

Reply:
It doesn't _have_ to, but it has to make some compromise between:

1. Defaulting to having a VERSION and SOVERSION
2. Enabling users to conveniently override "not defined"
3. Enabling users to conveniently override to "empty string"

I'm not willing to compromise on (1) because having library versions is the correct default for too many users. So then in order to override, the override variable must be defined. If it's defined, how should I specify that I should _remove_ the VERSION and SOVERSION properties, as I need to do for pip packaging? I can either:

1. Use `<empty>` as a special value leading to the empty string and let an empty string lead to undefined.
2. Use `<undef>` as a special value leading to undefined and let an empty string be an empty string.

The mathematician in me likes (2) because "an empty string [should] be an empty string". However, I chose (1) because of the vexing `..` behavior. I still haven't seen a _use_ for the empty string that isn't covered by `CMAKE_RELEASE_POSTFIX`. Well, I didn't quite choose (1) because I didn't implement `<empty>`, but again, I haven't seen a need.

You clearly have a reason, otherwise we wouldn't be in this discussion. Can you enlighten me?